SL tutorial draft

eddotman opened the pull request:
Draft tutorial for running pycc end to end and doing SL on a toy dataset where a synthetic function is used to create the initial dataset and update said dataset with additional "measured" candidates during SL.

@enze-chen -- I styled this after the existing tutorials but feel free to change/suggest whatever.

yoolheekim commented on api_examples/sequential_learning_api_tutorial.ipynb:
```diff
@@ -0,0 +1,692 @@
+   "source": [
+    "sl_results = {}\n",
+    "\n",
+    "best_sl_pred_vals, best_sl_measured_vals = run_design(\n",
```
change function name from `run_design` to `run_sequential_learning`

eddotman replied:
done :)

matSciMalcolm commented:
@eddotman Would you mind breaking out some of the functionality of `run_sequential_leanring` into individual functions so users can better understand how citrination-client.models opperates?

enze-chen commented on api_examples/sequential_learning_api_tutorial.ipynb:
```diff
@@ -0,0 +1,692 @@
+   ]
+  },
+  {
+   "cell_type": "code",
```
delete blank cell (Ctrl+Enter to run last cell, or just Cut every time)

enze-chen commented on api_examples/sequential_learning_api_tutorial.ipynb:
```diff
@@ -0,0 +1,692 @@
+    "\n",
+    "*Authors: Edward Kim, Enze Chen*\n",
+    "\n",
+    "In this notebook, we will cover how to perform **sequential learning** (SL) using the [Citrination API](http://citrineinformatics.github.io/python-citrination-client/). Sequential learning is the key workflow which allows machine learning algorithms and in-lab experiments to iteratively inform each other.\n",
```
would be nice to link to [sequential learning on citrine.io](https://citrine.io/platform/sequential-learning/), or another resource that elaborates more on this

enze-chen commented on api_examples/sequential_learning_api_tutorial.ipynb:
```diff
@@ -0,0 +1,692 @@
+    "\n",
+    "- Create and access datasets through the API ([documentation](http://citrineinformatics.github.io/python-citrination-client/tutorial/data_examples.html) and [tutorial](https://github.com/CitrineInformatics/community-tools/blob/master/api_examples/data_client_api_tutorial.ipynb))\n",
+    "- Create and access data views through the API ([tutorial](https://github.com/CitrineInformatics/community-tools/blob/master/api_examples/data_views_api_tutorial.ipynb))\n",
+    "- What the data views [front-end UI](https://citrination.com/data_views) looks like"
```
"looks like, particularly the Design interface." (to be as explicit/specific as possible)

enze-chen commented on api_examples/sequential_learning_api_tutorial.ipynb:
```diff
@@ -0,0 +1,692 @@
+   "source": [
+    "## BEFORE YOU RUN THIS NOTEBOOK\n",
+    "\n",
+    "This notebook uses some convenience functions to wrap several API endpoints. These are listed at the end of the notebook in the section titled **Convenience functions**. Run those functions first before running the main notebook!"
```
Nice, I like how this is at the end. Might be nice to create an internal link to Convenience function header with syntax: `[Text we read](#Header-name-with-spaces-hyphenated)` and then a link at the end of Convenience function taking you back up to Imports.

enze-chen commented on api_examples/sequential_learning_api_tutorial.ipynb:
```diff
@@ -0,0 +1,692 @@
+   "source": [
+    "# Plot initial training set, colored by toy function values\n",
+    "\n",
+    "plt.scatter(np.array(toy_x)[:,0], np.array(toy_x)[:,1],\n",
```
This is so pretty! <3 Can we add a colorbar? Optional, but for a second I wasn't sure which color was optimal (and not realizing that we're looking for points close to origin)

enze-chen commented on api_examples/sequential_learning_api_tutorial.ipynb:
```diff
@@ -0,0 +1,692 @@
+    "        _wait_on_data_view(client, dataset_id, view_id, wait_time, print_output)\n",
+    "            \n",
+    "        # Submit a design run\n",
+    "        design_id = client.submit_design_run(\n",
```
hmm this needs Admin privileges right? Not urgent for this demo, but how will we let external people do this step? Will have to check in with Kevin/Travis/Sean.

enze-chen commented on api_examples/sequential_learning_api_tutorial.ipynb:
```diff
@@ -0,0 +1,692 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Running design\n",
```
In this section, maybe also add a comment+internal link to read comments in `run_design()` so people know what they're actually doing (for those who care), because just this is pretty abstracted.

enze-chen commented on api_examples/sequential_learning_api_tutorial.ipynb:
```diff
@@ -0,0 +1,692 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sns.lineplot(\n",
```
I think I'm just being silly, but why does my output differ from plot for "best predicted value"? My "best predicted value" also never deviated from 14.43 between two runs, even when my "best initial value" in training set was 4.61. Should we be expecting strict improvement? Are these two metrics in fact not the same? It's 100% likely a misunderstanding on my part of what these metrics are.
![screen shot 2019-02-25 at 08 32 16](https://user-images.githubusercontent.com/20807339/53353747-46997980-38da-11e9-88fb-b9f8d9831835.png)

eddotman replied:
This confused me too for the longest time hahaha. It's because the plot shows the measured ground-truth results, and the output during the SL iterations shows the predicted values. i.e., you can do better/worse than predicted.

enze-chen commented on api_examples/sequential_learning_api_tutorial.ipynb:
```diff
@@ -0,0 +1,692 @@
+    ")\n",
+    "\n",
+    "# Reset dataset after each SL run\n",
+    "dataset_id = upload_data_and_get_id(\n",
```
A couple things:
1. Why do we run this one more time? A little more context would be nice.
1. For some reason none of the changes I made in the API is reflected in the UI. :( In terms of new PIFs and more rows in Data Views. Do you also observe this?

eddotman commented:
Thanks for the comments, everyone! I'll plan to address these early next week while also tackling some related comments from here https://github.com/CitrineInformatics/unobtainium/pull/2#pullrequestreview-205047480 

(Just to avoid having the code exist / be updated in two separate places)

eddotman commented:
Partially addressed some comments... still working on the rest through next week. Major update is on the output being a lot more informative at the end of this tutorial:
![image](https://user-images.githubusercontent.com/1232389/53668077-d9227b80-3c40-11e9-86ee-43675249ddfc.png)

enze-chen commented:
Ooooh, plots with uncertainty! <3 OK, thanks for correcting the legend. Also, in terms of breaking up the functions, I'd say you can lower the priority on that. Going off of your notebook, I went ahead and made a `models_client_api_tutorial.ipynb` to complement/expose other methods of `ModelsClient` (see [PR #11](https://github.com/CitrineInformatics/community-tools/pull/11)).

enze-chen commented:
RE Sale Demo: Could you add a `requirements.txt` file (maybe encompassing all Python packages in this tutorial; but just for your demo is enough) and add something like `pip install -U -r requirements.txt` to the README? I think this might help the setup go a little smoother for everyone in the future.

enze-chen commented:
I addressed most of my comments and added some style things, "lgtm" as Sean would say.   
Outstanding issues:
1. Logan mentioned moving functions into a `.py` file. Up to you if you want to hide the methods.
2. Still not sure how we will get non-admin users to run SL by themselves.

eddotman commented:
I think it's fine to leave the functions in the notebook, since we want this tutorial to be fully self-contained. 

Yeah, the admin account issue is something I'm not sure how to resolve. Since the tutorial is public, it would be nice to allow public users to run this code. I think it's fine for now to just note that admin accounts are needed on the public site.
